File: clients/eventodb_ex/test/read_test.exs
defmodule EventodbEx.ReadTest do
  use ExUnit.Case, async: true
  import EventodbEx.TestHelper

  setup do
    {client, namespace_id, _token} = create_test_namespace("read")

    on_exit(fn ->
      cleanup_namespace(client, namespace_id)
    end)

    %{client: client}
  end

  test "READ-001: Read from empty stream", %{client: client} do
    stream = unique_stream()

    assert {:ok, messages, _client} = EventodbEx.stream_get(client, stream)
    assert messages == []
  end

  test "READ-002: Read single message", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{foo: "bar"}}

    {:ok, _result, client} = EventodbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = EventodbEx.stream_get(client, stream)

    assert length(msg) == 7
    [id, type, position, global_position, data, metadata, time] = msg

    assert is_binary(id)
    assert type == "TestEvent"
    assert position == 0
    assert is_integer(global_position)
    assert data == %{"foo" => "bar"}
    assert is_map(metadata) or is_nil(metadata)
    assert is_binary(time)
  end

  test "READ-003: Read multiple messages", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{}}

    # Write 5 messages
    client =
      Enum.reduce(1..5, client, fn _, c ->
        {:ok, _result, new_c} = EventodbEx.stream_write(c, stream, message)
        new_c
      end)

    {:ok, messages, _client} = EventodbEx.stream_get(client, stream)
    assert length(messages) == 5

    # Verify positions are in order
    positions = Enum.map(messages, fn [_, _, pos, _, _, _, _] -> pos end)
    assert positions == [0, 1, 2, 3, 4]
  end

  test "READ-004: Read with position filter", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{}}

    # Write 10 messages
    client =
      Enum.reduce(1..10, client, fn _, c ->
        {:ok, _result, new_c} = EventodbEx.stream_write(c, stream, message)
        new_c
      end)

    {:ok, messages, _client} = EventodbEx.stream_get(client, stream, %{position: 5})
    assert length(messages) == 5

    positions = Enum.map(messages, fn [_, _, pos, _, _, _, _] -> pos end)
    assert positions == [5, 6, 7, 8, 9]
  end

  test "READ-005: Read with global position filter", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{}}

    # Write messages and track global positions
    results =
      Enum.reduce(1..4, {[], client}, fn _, {acc, c} ->
        {:ok, result, new_c} = EventodbEx.stream_write(c, stream, message)
        {[result | acc], new_c}
      end)
      |> elem(0)
      |> Enum.reverse()

    # Get the third message's global position
    target_gpos = Enum.at(results, 2).global_position

    {:ok, messages, _client} =
      EventodbEx.stream_get(client, stream, %{global_position: target_gpos})

    global_positions = Enum.map(messages, fn [_, _, _, gpos, _, _, _] -> gpos end)
    assert Enum.all?(global_positions, &(&1 >= target_gpos))
  end

  test "READ-006: Read with batch size limit", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{}}

    # Write 100 messages
    client =
      Enum.reduce(1..100, client, fn _, c ->
        {:ok, _result, new_c} = EventodbEx.stream_write(c, stream, message)
        new_c
      end)

    {:ok, messages, _client} = EventodbEx.stream_get(client, stream, %{batch_size: 10})
    assert length(messages) == 10

    positions = Enum.map(messages, fn [_, _, pos, _, _, _, _] -> pos end)
    assert positions == Enum.to_list(0..9)
  end

  test "READ-007: Read with batch size unlimited", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{}}

    # Write 50 messages
    client =
      Enum.reduce(1..50, client, fn _, c ->
        {:ok, _result, new_c} = EventodbEx.stream_write(c, stream, message)
        new_c
      end)

    {:ok, messages, _client} = EventodbEx.stream_get(client, stream, %{batch_size: -1})
    assert length(messages) == 50
  end

  test "READ-008: Read message data integrity", %{client: client} do
    stream = unique_stream()

    complex_data = %{
      nested: %{
        array: [1, 2, 3],
        bool: true,
        null: nil
      }
    }

    message = %{type: "TestEvent", data: complex_data}

    {:ok, _result, client} = EventodbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = EventodbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, data, _metadata, _time] = msg

    # Verify deep equality
    assert data["nested"]["array"] == [1, 2, 3]
    assert data["nested"]["bool"] == true
    assert data["nested"]["null"] == nil
  end

  test "READ-009: Read message metadata integrity", %{client: client} do
    stream = unique_stream()

    metadata = %{
      correlationId: "123",
      userId: "user-456"
    }

    message = %{type: "TestEvent", data: %{}, metadata: metadata}

    {:ok, _result, client} = EventodbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = EventodbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, _data, read_metadata, _time] = msg

    assert read_metadata["correlationId"] == "123"
    assert read_metadata["userId"] == "user-456"
  end

  test "READ-010: Read message timestamp format", %{client: client} do
    stream = unique_stream()
    message = %{type: "TestEvent", data: %{}}

    {:ok, _result, client} = EventodbEx.stream_write(client, stream, message)
    {:ok, [msg], _client} = EventodbEx.stream_get(client, stream)

    [_id, _type, _pos, _gpos, _data, _metadata, time] = msg

    # Verify ISO 8601 timestamp format (with or without nanoseconds)
    assert String.match?(time, ~r/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$/)
  end
end

File: clients/messagedb-node/tests/read.test.ts
import { describe, test, expect, afterEach } from 'vitest';
import { setupTest, randomStreamName, type TestContext } from './helpers.js';

describe('READ Tests', () => {
  const contexts: TestContext[] = [];

  afterEach(async () => {
    await Promise.all(contexts.map(ctx => ctx.cleanup()));
    contexts.length = 0;
  });

  test('READ-001: Read from empty stream', async () => {
    const ctx = await setupTest('read-001');
    contexts.push(ctx);

    const stream = randomStreamName();
    const messages = await ctx.client.streamGet(stream);

    expect(messages).toEqual([]);
  });

  test('READ-002: Read single message', async () => {
    const ctx = await setupTest('read-002');
    contexts.push(ctx);

    const stream = randomStreamName();
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { foo: 'bar' }
    });

    const messages = await ctx.client.streamGet(stream);
    expect(messages).toHaveLength(1);

    const [id, type, position, globalPosition, data, metadata, time] = messages[0];
    expect(typeof id).toBe('string');
    expect(type).toBe('TestEvent');
    expect(position).toBe(0);
    expect(typeof globalPosition).toBe('number');
    expect(data).toEqual({ foo: 'bar' });
    expect(typeof time).toBe('string');
  });

  test('READ-003: Read multiple messages', async () => {
    const ctx = await setupTest('read-003');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write 5 messages
    for (let i = 0; i < 5; i++) {
      await ctx.client.streamWrite(stream, {
        type: 'Event',
        data: { n: i }
      });
    }

    const messages = await ctx.client.streamGet(stream);
    expect(messages).toHaveLength(5);

    // Check positions are in order
    for (let i = 0; i < 5; i++) {
      const [, , position] = messages[i];
      expect(position).toBe(i);
    }
  });

  test('READ-004: Read with position filter', async () => {
    const ctx = await setupTest('read-004');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write 10 messages
    for (let i = 0; i < 10; i++) {
      await ctx.client.streamWrite(stream, {
        type: 'Event',
        data: { n: i }
      });
    }

    const messages = await ctx.client.streamGet(stream, { position: 5 });
    expect(messages).toHaveLength(5);

    // Check we got positions 5-9
    expect(messages[0][2]).toBe(5);
    expect(messages[4][2]).toBe(9);
  });

  test('READ-005: Read with global position filter', async () => {
    const ctx = await setupTest('read-005');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write 4 messages
    const results = [];
    for (let i = 0; i < 4; i++) {
      const result = await ctx.client.streamWrite(stream, {
        type: 'Event',
        data: { n: i }
      });
      results.push(result);
    }

    // Read from global position of first message
    const targetGlobalPos = results[0].globalPosition;
    const messages = await ctx.client.streamGet(stream, { 
      globalPosition: targetGlobalPos 
    });

    // Should get all 4 messages
    expect(messages.length).toBeGreaterThanOrEqual(1);
    // All returned messages should have global positions in correct range
    const firstMsgGlobalPos = messages[0][3];
    expect(firstMsgGlobalPos).toBeGreaterThanOrEqual(targetGlobalPos);
  });

  test('READ-006: Read with batch size limit', async () => {
    const ctx = await setupTest('read-006');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write 100 messages
    for (let i = 0; i < 100; i++) {
      await ctx.client.streamWrite(stream, {
        type: 'Event',
        data: { n: i }
      });
    }

    const messages = await ctx.client.streamGet(stream, { batchSize: 10 });
    expect(messages).toHaveLength(10);
    
    // Should get positions 0-9
    expect(messages[0][2]).toBe(0);
    expect(messages[9][2]).toBe(9);
  });

  test('READ-007: Read with batch size unlimited', async () => {
    const ctx = await setupTest('read-007');
    contexts.push(ctx);

    const stream = randomStreamName();
    
    // Write 50 messages
    for (let i = 0; i < 50; i++) {
      await ctx.client.streamWrite(stream, {
        type: 'Event',
        data: { n: i }
      });
    }

    const messages = await ctx.client.streamGet(stream, { batchSize: -1 });
    expect(messages).toHaveLength(50);
  });

  test('READ-008: Read message data integrity', async () => {
    const ctx = await setupTest('read-008');
    contexts.push(ctx);

    const stream = randomStreamName();
    const complexData = {
      nested: {
        array: [1, 2, 3],
        bool: true,
        nullValue: null
      }
    };

    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: complexData
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , data] = messages[0];
    expect(data).toEqual(complexData);
  });

  test('READ-009: Read message metadata integrity', async () => {
    const ctx = await setupTest('read-009');
    contexts.push(ctx);

    const stream = randomStreamName();
    const metadata = {
      correlationId: '123',
      userId: 'user-456'
    };

    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { test: true },
      metadata
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , , readMetadata] = messages[0];
    expect(readMetadata).toEqual(metadata);
  });

  test('READ-010: Read message timestamp format', async () => {
    const ctx = await setupTest('read-010');
    contexts.push(ctx);

    const stream = randomStreamName();
    await ctx.client.streamWrite(stream, {
      type: 'TestEvent',
      data: { test: true }
    });

    const messages = await ctx.client.streamGet(stream);
    const [, , , , , , time] = messages[0];
    
    // Should be valid ISO 8601 format
    expect(typeof time).toBe('string');
    
    // Verify it's a valid date
    const parsed = new Date(time);
    expect(parsed.toString()).not.toBe('Invalid Date');
    
    // Should match pattern YYYY-MM-DDTHH:MM:SS.nnnZ or similar
    expect(time).toMatch(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$/);
  });
});
